### ISSUE TITLE
release in maven central

### ISSUE THREAD

**Issue by stokpop** [Cryptolens/cryptolens-java#20]
Hi,

to help with our packaging of this library it would really help if it got released in maven central.

Is it possible for you to release to maven central?

Thanks.

**Comment by artemlos** [comment]
@stokpop Hi! We are working on adding it to the maven central repository. I will let you know once it is available. /Artem

**Comment by stokpop** [comment]
@artemlos Hi Artem, is there progress to report on this issue? Thanks.

**Comment by artemlos** [comment]
@stokpop Hi! Sorry, it took longer than expected. We will try to get it done by the end of next week.

**Comment by sergey-oplavin** [comment]
Hi Artem, just wanted to ask if there is any update on the issue? 
Adding cryptolens jar to maven/gradle project is not very convenient now, so it would be really nice to have it published into Maven Central (or to any other place).  
We have recently published one of our public packages to Maven Central, so if you have any questions about the procedure - don't hesitate to ask. Thanks.

**Comment by artemlos** [comment]
@sergey-oplavin thanks for offering your help!

I think we got stuck on either creating the correct configuration. There is already a package registered on Sonatype (https://issues.sonatype.org/browse/OSSRH-61669), which I believe was the first step that was necessary.

I will look into it and get back where exactly we had issues. We use IntelliJ, so I will look for a tutorial and try it again.

**Comment by artemlos** [comment]
@sergey-oplavin Just wanted to check with you if you recognize the following authentication error when running mvn deloy? https://s01.oss.sonatype.org/content/repositories/snapshots/io/cryptolens/cryptolens/1.24-SNAPSHOT/cryptolens-1.24-20211117.134324-1.pom.asc

I will look into it more, but I just wanted to check in case you have seen this error before and know the possible cause.

**Comment by sergey-oplavin** [comment]
@artemlos Did you provide your username/password for sonatype server in settings.xml?

**Comment by artemlos** [comment]
@sergey-oplavin Yep, that was done. I followed the following tutorial: https://central.sonatype.org/publish/publish-maven/#

I will double check that the folder is correct, etc. It might be that IntelliJ does not find the file.

**Comment by sergey-oplavin** [comment]
@artemlos Just verified - if I don't have the following section in my ~/.m2/settings.xml file, it shows the same error on deploy.
```xml
<server>
	<id>ossrh</id>   
	<username>user</username>
	<password>pwd</password>
</server>
```
Where user and pwd are credentials for sonatype (https://s01.oss.sonatype.org/)

**Comment by artemlos** [comment]
@sergey-oplavin I have the configuration that you listed above, but it appears that when I authenticate on https://s01.oss.sonatype.org/, I get an authentication error:

![image](https://user-images.githubusercontent.com/6392760/142252216-dd313fd4-f9b8-4858-af94-13e4de7dbddb.png)

I will look into how new credentials can be created for that page. I thought that I could just use the same as for Sonatype Jira.

**Comment by sbraconnier** [comment]
Hello,

We are evaluating the cryptolens software licensing system for our product, but came across this issue. It's been 4 years since the last issue update. Any news about this issue??

Note that the OSSRH service has reached end-of-life on june 25, 2025

You must now publish through the Central Publishing Portal:

https://central.sonatype.org/register/central-portal/

I personally own an open source project published in the maven central and can tell you that the new portal is easier to setup and use.

Thanks.

**Comment by artemlos** [comment]
@sbraconnier 

Thank you for the feedback! I'm sorry for a late reply.

We are working to remove OSSRH dependency. In meantime, it's possible to use cryptolens-android.jar, a build that does not include it. It contains the same methods except for OSSRH dependency.  We needed OSSRH to generate the machine code, however, it's possible to use other ways to identify a machine that do not rely on OSSRH .

Thank you for mentioning the new portal, we'll take a look and try to publish again.

Thanks!

**Comment by sbraconnier** [comment]
@artemlos Sorry if I'm being insistent but I'm trying to fully test cryptolens within the trial period :).

I'll try to be more clear regarding our setup.

We have a product that already ships the `gson`, `apache-codec` and `oshi` libraries. So when we are introducing a new library (like the cryptolens one), we usually exclude the dependencies we already provide, which is possible to do when the artifact is published into the maven repo. Here's a example (using gradle kotlin DSL) that exclude the slf4j library:

```
implementation("com.zaxxer:HikariCP:4.0.3") {
    exclude(group = "org.slf4j")
}
```

If we include fat jars in our project (a jar that includes its dependencies in the same jar), it would then be impossible for us to guarantee which version will be used for a library that is included into those fat jars. Usually, when you create a java library to be used by others in their project, you provide a jar that doesn't include any other dependencies, as they can come into conflict with other versions of the same library in the whole project. We cannot guarantee which class the JVM classloader will load first.

Suppose the gson 2.9.1 library has a vulnerability issue, it would be impossible for us to just update the gson library in our project, since we wouldn't be able to guarantee that the classes from the updated library would be taken over the ones in the cryptolens fat jar.

> It may be possible to ship a version of the library with no dependencies, and expose an abstract method for e.g. json parsing, which would enable you to then to use any JSON library and we would then be able to drop the dependency on gson. I think this should work in Java. We will explore this option too.

You don't have to expose abstract methods for json parsing. You can just document that a gson library must be present in the classpath in order for your library to work, specifying which version was used for the build. The user of your library will be able to choose which version he wants to use, and will be responsible to read the release notes of the gson libraries to see if some breaking changes may cause problems (which is very rare for this kind of library).

Actually, we are using the `cryptolens-android` jar, which doesn't include the `oshi` library, but we still use the `Helpers.getMachineCode()`, because the `oshi` library is present in the classpath (provided by us).

By the way, the documentation is confusing. From you website :

> There are two pre-compiled jar files available in the [package repository](https://github.com/Cryptolens/cryptolens-java): `cryptolens.jar` and `cryptolens-android.jar`. If your application is cross platform or if you would like to have as few dependencies as possible (e.g., without `slf4j`), we recommend to use `cryptolens-android.jar` instead.
> If you choose to use cryptolens-android.jar, GetMachineCode and IsOnRightMachine need to be called with the version parameter set to 2. For example, `Helpers.GetMachineCode(2)` or `Helpers.IsOnRightMachine(license, 2)`

But then using `Helpers.GetMachineCode(2)` only supports Windows. See the [implementation](https://github.com/Cryptolens/cryptolens-java/blob/master/src/main/java/io/cryptolens/methods/Helpers.java#L56). So it is not  
cross platform at all. In our case, we can't use `Helpers.GetMachineCode(2)` because we must support more than just Windows.

> For now, it appears maven approach would be the easiest

Indeed.

> although I'm not sure if there can be any problems in the future if they release breaking changes to one of our dependencies

Even using the fat jar you provide, we can't know for sure that there wouldn't be any problems. As soon as you have more that 1 version of the same class in the classpath, it could be problematic.

**Comment by artemlos** [comment]
@sbraconnier Thank you for a very detailed answer and tips. I really appreciate it. 

I wanted to address some of the questions first and get back later with a reply to the other suggestions.

> But then using Helpers.GetMachineCode(2) only supports Windows. See the [implementation](https://github.com/Cryptolens/cryptolens-java/blob/master/src/main/java/io/cryptolens/methods/Helpers.java#L56). So it is not
cross platform at all. In our case, we can't use Helpers.GetMachineCode(2) because we must support more than just Windows.

An update was released that supports other platforms too, not only Windows: https://github.com/Cryptolens/cryptolens-java/commit/f3adb600ddce85926fc740818fe5de047cabb57f

> Sorry if I'm being insistent but I'm trying to fully test cryptolens within the trial period :).
No worries :) if you need a longer trial, please let us know at support@cryptolens.io.

**Comment by artemlos** [comment]
@sbraconnier @sergey-oplavin @stokpop The package should now be available through maven central.

A few changes since last release:
1. We dropped commons-codec and replaced with a built in library in Java 8. Java 7 would require it but it appears it is deprecated a long time ago.
2. GetMachineCode is cross platform. It's implemented based on the documented that was shared earlier. We need to test it on Mac and Linux, but on Windows it works flawlessly based on testing.
3. OSSH is option. To avoid it, you can just call GetMachineCode with v=2 (Win) or v=3(Win+Mac+Linux).
4. If you use OSSH, we fixed some issues that caused issues on newer systems (another client emailed us about it).

https://central.sonatype.com/artifact/io.cryptolens/cryptolens

**Comment by artemlos** [comment]
Thank you! We will document this more thoroughly in the coming days. 

Correct, it was the OSHI I ment initially. Sorry for the confusion  :)

**Comment by artemlos** [comment]
Solved in https://github.com/Cryptolens/cryptolens-java/releases/tag/v1.27
